Build failure (C4996) when including opentelemetry/logs/provider.h with opentelemetry-cpp[geneva]

## Steps to Reproduce
1. Follow this article to integrate vcpkg with MSBuild/VisualStudio.

     Tutorial: Install and use packages with MSBuild in Visual Studio
     https://learn.microsoft.com/en-us/vcpkg/get_started/get-started-msbuild

     ```sh
     .\vcpkg.exe integrate install
     ```
2. Create a C++ console application in Visual Studio 2022.
3. In Visual Studio, open Tools → Command Line → Developer Command Prompt.
4. Run:
    ```sh
   vcpkg new --application
   vcpkg add port opentelemetry-cpp[geneva]
   ```
5. In the project Property Pages, set:   vcpkg → Use Vcpkg Manifest = Yes
6. Include the following header in your .cpp file:
    ```C++
   #include "opentelemetry/logs/provider.h"
   ```
7. Build the project.

## Expected Result
The project builds successfully with no errors.

## Actual Result
The build fails with error C4996 due to deprecated symbols:

```text
opentelemetry::v1::logs::EventLogger was declared deprecated
opentelemetry::v1::logs::EventLogger::~EventLogger was declared deprecated
opentelemetry::v1::logs::EventLoggerProvider was declared deprecated
opentelemetry::v1::logs::EventLoggerProvider::~EventLoggerProvider was declared deprecated
opentelemetry::v1::logs::Provider::GetEventProvider was declared deprecated
```
## Additional Notes
This header is included because its APIs are required to set the Geneva `TracerProvider` as the global default, specifically to call:

GetLoggerProvider()
SetLoggerProvider()

Here is my testing project for this issue: https://github.com/lukeina2z/playground/tree/main/cpp/LogProviderWarning

--- a/api/include/opentelemetry/logs/event_logger.h
+++ b/api/include/opentelemetry/logs/event_logger.h
@@ -15,6 +15,17 @@ OPENTELEMETRY_BEGIN_NAMESPACE
 namespace logs
 {
 #if OPENTELEMETRY_ABI_VERSION_NO < 2
+#  if defined(_MSC_VER)
+#    pragma warning(push)
+#    pragma warning(disable : 4996)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#  endif
+
 /**
  * Handles event log record creation.
  **/
@@ -77,6 +88,14 @@ private:
   void IgnoreTraitResult(ValueType &&...)
   {}
 };
+
+#  if defined(_MSC_VER)
+#    pragma warning(pop)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic pop
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic pop
+#  endif
 #endif
 }  // namespace logs
 OPENTELEMETRY_END_NAMESPACE

--- a/api/include/opentelemetry/logs/event_logger_provider.h
+++ b/api/include/opentelemetry/logs/event_logger_provider.h
@@ -15,6 +15,17 @@ class EventLogger;
 class Logger;
 
 #if OPENTELEMETRY_ABI_VERSION_NO < 2
+#  if defined(_MSC_VER)
+#    pragma warning(push)
+#    pragma warning(disable : 4996)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#  endif
+
 /**
  * Creates new EventLogger instances.
  */
@@ -32,6 +43,14 @@ public:
       nostd::shared_ptr<Logger> delegate_logger,
       nostd::string_view event_domain) noexcept = 0;
 };
+
+#  if defined(_MSC_VER)
+#    pragma warning(pop)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic pop
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic pop
+#  endif
 #endif
 }  // namespace logs
 OPENTELEMETRY_END_NAMESPACE

--- a/api/include/opentelemetry/logs/noop.h
+++ b/api/include/opentelemetry/logs/noop.h
@@ -91,10 +91,22 @@ private:
 };
 
 #if OPENTELEMETRY_ABI_VERSION_NO < 2
+#  if defined(_MSC_VER)
+#    pragma warning(push)
+#    pragma warning(disable : 4996)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic warning "-Wdeprecated-declarations"
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic push
+#    pragma clang diagnostic warning "-Wdeprecated-declarations"
+#  endif
+
 class NoopEventLogger final : public EventLogger
 {
 public:
   NoopEventLogger() : logger_{nostd::shared_ptr<NoopLogger>(new NoopLogger())} {}
+  ~NoopEventLogger() override = default;
 
   const nostd::string_view GetName() noexcept override { return "noop event logger"; }
 
@@ -112,27 +124,9 @@ private:
 class NoopEventLoggerProvider final : public EventLoggerProvider
 {
 public:
-#  if defined(_MSC_VER)
-#    pragma warning(push)
-#    pragma warning(disable : 4996)
-#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
-#    pragma GCC diagnostic push
-#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-#  elif defined(__clang__) || defined(__apple_build_version__)
-#    pragma clang diagnostic push
-#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
-#  endif
-
   NoopEventLoggerProvider() : event_logger_{nostd::shared_ptr<EventLogger>(new NoopEventLogger())}
   {}
-
-#  if defined(_MSC_VER)
-#    pragma warning(pop)
-#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
-#    pragma GCC diagnostic pop
-#  elif defined(__clang__) || defined(__apple_build_version__)
-#    pragma clang diagnostic pop
-#  endif
+  ~NoopEventLoggerProvider() override = default;
 
   nostd::shared_ptr<EventLogger> CreateEventLogger(
       nostd::shared_ptr<Logger> /*delegate_logger*/,
@@ -144,6 +138,14 @@ public:
 private:
   nostd::shared_ptr<EventLogger> event_logger_;
 };
+
+#  if defined(_MSC_VER)
+#    pragma warning(pop)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic pop
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic pop
+#  endif
 #endif
 
 }  // namespace logs

--- a/api/include/opentelemetry/logs/noop.h
+++ b/api/include/opentelemetry/logs/noop.h
@@ -96,10 +96,10 @@ private:
 #    pragma warning(disable : 4996)
 #  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
 #    pragma GCC diagnostic push
-#    pragma GCC diagnostic warning "-Wdeprecated-declarations"
+#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #  elif defined(__clang__) || defined(__apple_build_version__)
 #    pragma clang diagnostic push
-#    pragma clang diagnostic warning "-Wdeprecated-declarations"
+#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
 #  endif
 
 class NoopEventLogger final : public EventLogger

--- a/api/include/opentelemetry/logs/provider.h
+++ b/api/include/opentelemetry/logs/provider.h
@@ -53,10 +53,10 @@ public:
 #    pragma warning(disable : 4996)
 #  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
 #    pragma GCC diagnostic push
-#    pragma GCC diagnostic warning "-Wdeprecated-declarations"
+#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #  elif defined(__clang__) || defined(__apple_build_version__)
 #    pragma clang diagnostic push
-#    pragma clang diagnostic warning "-Wdeprecated-declarations"
+#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
 #  endif
   /**
    * Returns the singleton EventLoggerProvider.
@@ -102,10 +102,10 @@ private:
 #    pragma warning(disable : 4996)
 #  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
 #    pragma GCC diagnostic push
-#    pragma GCC diagnostic warning "-Wdeprecated-declarations"
+#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
 #  elif defined(__clang__) || defined(__apple_build_version__)
 #    pragma clang diagnostic push
-#    pragma clang diagnostic warning "-Wdeprecated-declarations"
+#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
 #  endif
 
   OPENTELEMETRY_DEPRECATED

--- a/api/include/opentelemetry/logs/provider.h
+++ b/api/include/opentelemetry/logs/provider.h
@@ -48,6 +48,16 @@ public:
   }
 
 #if OPENTELEMETRY_ABI_VERSION_NO < 2
+#  if defined(_MSC_VER)
+#    pragma warning(push)
+#    pragma warning(disable : 4996)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic push
+#    pragma GCC diagnostic warning "-Wdeprecated-declarations"
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic push
+#    pragma clang diagnostic warning "-Wdeprecated-declarations"
+#  endif
   /**
    * Returns the singleton EventLoggerProvider.
    *
@@ -70,6 +80,13 @@ public:
     std::lock_guard<common::SpinLockMutex> guard(GetLock());
     GetEventProvider() = tp;
   }
+#  if defined(_MSC_VER)
+#    pragma warning(pop)
+#  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
+#    pragma GCC diagnostic pop
+#  elif defined(__clang__) || defined(__apple_build_version__)
+#    pragma clang diagnostic pop
+#  endif
 #endif
 
 private:
@@ -80,23 +97,24 @@ private:
   }
 
 #if OPENTELEMETRY_ABI_VERSION_NO < 2
-  OPENTELEMETRY_DEPRECATED
-  OPENTELEMETRY_API_SINGLETON static nostd::shared_ptr<EventLoggerProvider> &
-  GetEventProvider() noexcept
-  {
 #  if defined(_MSC_VER)
 #    pragma warning(push)
 #    pragma warning(disable : 4996)
 #  elif defined(__GNUC__) && !defined(__clang__) && !defined(__apple_build_version__)
 #    pragma GCC diagnostic push
-#    pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#    pragma GCC diagnostic warning "-Wdeprecated-declarations"
 #  elif defined(__clang__) || defined(__apple_build_version__)
 #    pragma clang diagnostic push
-#    pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#    pragma clang diagnostic warning "-Wdeprecated-declarations"
 #  endif
 
+  OPENTELEMETRY_DEPRECATED
+  OPENTELEMETRY_API_SINGLETON static nostd::shared_ptr<EventLoggerProvider> &
+  GetEventProvider() noexcept
+  {
     static nostd::shared_ptr<EventLoggerProvider> provider(new NoopEventLoggerProvider);
     return provider;
+  }
 
 #  if defined(_MSC_VER)
 #    pragma warning(pop)
@@ -105,7 +123,6 @@ private:
 #  elif defined(__clang__) || defined(__apple_build_version__)
 #    pragma clang diagnostic pop
 #  endif
-  }
 #endif
 
   OPENTELEMETRY_API_SINGLETON static common::SpinLockMutex &GetLock() noexcept

--- a/examples/common/logs_foo_library/CMakeLists.txt
+++ b/examples/common/logs_foo_library/CMakeLists.txt
@@ -12,3 +12,24 @@ target_include_directories(common_logs_foo_library
                            PUBLIC $<BUILD_INTERFACE:${EXAMPLES_COMMON_DIR}>)
 
 target_link_libraries(common_logs_foo_library PUBLIC opentelemetry-cpp::api)
+
+# Internal deprecated API calling should not failed the build in maintainer mode
+if(OTELCPP_MAINTAINER_MODE)
+  if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang|AppleClang")
+    get_target_property(COMMON_LOGS_FOO_COMPILE_OPTS common_logs_foo_library
+                        COMPILE_OPTIONS)
+    list(REMOVE_ITEM COMMON_LOGS_FOO_COMPILE_OPTS
+         -Wno-error=deprecated-declarations)
+    set_target_properties(
+      common_logs_foo_library PROPERTIES COMPILE_OPTIONS
+                                         "${COMMON_LOGS_FOO_COMPILE_OPTS}")
+  elseif(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    get_target_property(COMMON_LOGS_FOO_COMPILE_OPTS common_logs_foo_library
+                        COMPILE_OPTIONS)
+    list(REMOVE_ITEM COMMON_LOGS_FOO_COMPILE_OPTS /wd4996)
+    list(APPEND COMMON_LOGS_FOO_COMPILE_OPTS /w34996)
+    set_target_properties(
+      common_logs_foo_library PROPERTIES COMPILE_OPTIONS
+                                         "${COMMON_LOGS_FOO_COMPILE_OPTS}")
+  endif()
+endif()

